src/language/patterns.md

```diff
@@ -0,0 +1,358 @@
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b); // Swap.
+print('$a $b'); // Prints "right left".
+```
+
+### Switch statements and expressions
+
+Every case clause contains a pattern. This applies to [switch statements][switch]
+and [expressions][], as well as [if-case statements][if].
+
+_Case patterns_ are [refutable][];
+they allow control flow to either match and destructure the object being switched
+on, or continue execution if the object doesn't match. Case patterns can be
+[any kind of pattern][types]:
+
+```dart
+switch (obj) {
+  case 1: ...
+    // Matches if 1 == obj.
+  case >= first && <= last: ...  
+    // Matches if the value of obj is between the constant values of 'first' and 'last'.
+  case (var a, var b): ...
+    // Matches if obj is a record with two fields, then assigns the fields to 'a' and 'b'.
+  case Rect(width: var w, height: var h): ...
+    // Matches if obj is an instance of Rect, and then against the properties of Rect.
+  default: ...
+}
+```
+
+
+Values destructured by a pattern in a case become local variables,
+only available in the body of that case. 
+
+### For and for-in loops
+
+You can use patterns in [for and for-in loops][for] to iterate-over and destructure
+values in a collection.
+
+This example uses [object destructuring][object] in a for-in loop to destructure
+the [`MapEntry`][] objects that a `<Map>.entries` call returns:
+
+```dart
+Map<String, int> hist = // ...
+
+for (var MapEntry(key: key, value: count) in hist.entries) {
+  print('$key occurred $count times');
+}
+```
+
+The object pattern checks that `hist.entries` has the named type `MapEntry`,
+and then recurses into the named field subpatterns `key` and `value`.
+It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
+and binds the results to local variables `key` and `count`, respectively.
+
+Binding the result of a getter call to a variable of the same name is a common
+use case, so object patterns can also infer the getter name from the
+[variable subpattern][variable]. This allows you to simplify the variable pattern
+from something redundant like `key: key` to just `:key`:
+
+```dart
+for (var MapEntry(:key, value: count) in hist.entries) { 
+  print('$key occurred $count times');
+}
+```
+
+## Uses cases for patterns
+
+The [section above](#places-patterns-can-appear)
+describes _how_ you can use patterns in Dart code constructs, 
+sometimes with interesting uses cases as examples (like [swapping](#variable-assignment)
+the values of two variables, or
+[destructuring key-value pairs](#for-and-for-in-loops)
+in a map). 
+
+This section describes even more use cases, answering _when and why_ you might
+want to use patterns, the kinds of problems they solve, and the idioms they're 
+best suited for.
```

Comment:
**issue:** 
Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
This section describes even more use cases that answer:
* _when and why_ you might want to use patterns.
* the kinds of problems they solve.
* the idioms for which they're best suited.
```